sql/src/main/java/io/druid/sql/avatica/DruidMeta.java

```diff
@@ -0,0 +1,590 @@
+
+    return sqlResultSet(ch, sql);
+  }
+
+  private DruidConnection getDruidConnection(final String connectionId)
+  {
+    return getDruidConnection(connectionId, false);
+  }
+
+  private DruidConnection getDruidConnection(final String connectionId, final boolean createIfNotExists)
+  {
+    DruidConnection connection;
+
+    synchronized (connections) {
+      connection = connections.get(connectionId);
+
+      if (connection == null && createIfNotExists) {
+        if (connections.size() >= config.getMaxConnections()) {
+          throw new ISE("Too many connections, limit is[%,d]", config.getMaxConnections());
+        }
+        connection = new DruidConnection();
+        connections.put(connectionId, connection);
+        log.debug("Connection[%s] opened.", connectionId);
+      }
+
+      if (connection == null) {
+        throw new ISE("Connection[%s] not open", connectionId);
+      }
+    }
+
+    final DruidConnection finalConnection = connection;
+
+    return finalConnection.sync(
+        exec.schedule(
+            new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                final List<DruidStatement> statements = new ArrayList<>();
+
+                synchronized (connections) {
+                  if (connections.remove(connectionId) == finalConnection) {
+                    statements.addAll(finalConnection.statements().values());
+                    log.debug("Connection[%s] timed out, closing %,d statements.", connectionId, statements.size());
+                  }
+                }
+
+                final Closer closer = Closer.create();
+                for (final DruidStatement statement : statements) {
+                  closer.register(statement);
+                }
+                try {
+                  closer.close();
+                }
+                catch (IOException e) {
+                  throw Throwables.propagate(e);
+                }
+              }
+            },
+            new Interval(new DateTime(), config.getConnectionIdleTimeout()).toDurationMillis(),
+            TimeUnit.MILLISECONDS
+        )
+    );
+  }
+
+  private DruidStatement getDruidStatement(final StatementHandle statement)
+  {
+    synchronized (connections) {
+      final DruidConnection connection = getDruidConnection(statement.connectionId);
+      final DruidStatement druidStatement = connection.statements().get(statement.id);
+      Preconditions.checkState(druidStatement != null, "Statement[%s] does not exist", statement.id);
+      return druidStatement;
+    }
+  }
+
+  private void closeDruidStatement(final DruidStatement statement)
+  {
+    final DruidStatement druidStatement2;
+    synchronized (connections) {
+      final DruidConnection connection = getDruidConnection(statement.getConnectionId());
+      druidStatement2 = connection.statements().remove(statement.getStatementId());
+    }
+
+    if (statement == druidStatement2) {
```

Comment:
It would be extremely unlikely, but it could happen if the original statement was closed, then another was opened with the same id. That could only happen if the id rolled over Integer.MAX_VALUE in the meantime. Really unlikely but could happen in principle.

I can add a log message here that something funky may be going on.